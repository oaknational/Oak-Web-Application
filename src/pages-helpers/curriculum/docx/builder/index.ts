export { default as frontCover } from "./1_frontCover";
export { default as tableOfContents } from "./2_tableOfContents";
export { default as ourCurriculum } from "./3_ourCurriculum";
export { default as threadsExplainer } from "./4_threadsExplainer";
export { default as subjectExplainer } from "./5_subjectExplainer";
export { default as subjectPrincipals } from "./6_subjectPrincipals";
export { default as ourPartner } from "./7_ourPartner";
export { default as units } from "./8_units";
export { default as threadsOverview } from "./9_threadsOverview";
export { default as threadsDetail } from "./10_threadsDetail";
export { default as backCover } from "./11_backCover";
export { default as pageLayout } from "./12_pageLayout";
