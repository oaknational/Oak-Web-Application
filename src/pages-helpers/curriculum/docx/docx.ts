import { extname, join, relative } from "path";
import { readFile, stat } from "fs/promises";
import { createHash } from "crypto";

import { glob } from "glob";
import JSZip from "jszip";
import { ElementCompact, type Element } from "xml-js";

import {
  collapseFragments,
  jsonXmlToXmlString,
  safeXml,
  xmlElementToJson,
  xmlRootToJson,
} from "./xml";

function generateHash(buffer: Buffer | string) {
  const hash = createHash("SHA-256");
  hash.setEncoding("hex");
  hash.write(buffer);
  hash.end();
  return hash.read();
}

let maxId = 10000;
/**
 * Generates a <w:abstractNum/> / <w:num/> pair
 * @param zip docx zip file
 * @param numberingDefinition key / definition rules
 * @returns key / id lookup table
 */
export async function insertNumbering<T extends Record<string, string>>(
  zip: JSZipCached,
  numberingDefinition: T,
) {
  const lookup: Record<keyof typeof numberingDefinition, string> = {} as Record<
    keyof typeof numberingDefinition,
    string
  >;

  const docNumberingPath = "word/numbering.xml";
  const json = await zip.readJson(docNumberingPath);

  json.elements[0].elements = json.elements[0].elements ?? [];

  for (const [key, definition] of Object.entries(numberingDefinition)) {
    const abstractNumId = maxId++;
    const numId = maxId++;
    lookup[key as keyof T] = String(numId);
    json.elements[0].elements.push(
      xmlElementToJson(safeXml`
        <w:abstractNum w:abstractNumId="${abstractNumId}">
          ${definition}
        </w:abstractNum>
      `),
    );

    json.elements[0].elements.push(
      xmlElementToJson(safeXml`
        <w:num w:numId="${numId}">
          <w:abstractNumId w:val="${abstractNumId}" />
        </w:num>
      `),
    );
  }

  zip.writeJson(docNumberingPath, collapseFragments(json as Element));

  return lookup;
}

function extnameWithoutQuery(filepath: string) {
  return extname(filepath).replace(/\?.*$/, "");
}

export async function insertImages<T extends Record<string, string>>(
  zip: JSZipCached,
  images: T,
) {
  const DOC_RELS = "word/_rels/document.xml.rels";
  const json = await zip.readJson(DOC_RELS);

  const existingImageIds = json.elements[0].elements
    .map((element: Element) => element.attributes?.Id)
    .filter(Boolean);

  const output: Record<keyof typeof images, string> = {} as Record<
    keyof typeof images,
    string
  >;
  const elements = await Promise.all(
    Object.entries(images).map(
      async ([key, filePathOrUrl]: [keyof typeof images, string]) => {
        const imagePathHash = generateHash(filePathOrUrl);
        const id = "rId" + imagePathHash;
        output[key] = id;

        if (existingImageIds.includes(id)) {
          return null;
        } else {
          existingImageIds.push(id);
        }

        let ext: string;
        let file: Buffer;
        if (filePathOrUrl.match(/^(data):/)) {
          file = Buffer.from(await (await fetch(filePathOrUrl)).arrayBuffer());
          ext =
            filePathOrUrl.match(/^data:image\/([^;]+);base64/)?.[1] ??
            "unknown";
        } else if (filePathOrUrl.match(/^(https?):/)) {
          const res = await fetch(filePathOrUrl);
          ext =
            "." + res.headers.get("content-type")?.split("/")[1] ?? "unknown";
          file = Buffer.from(await res.arrayBuffer());
          // ext = extnameWithoutQuery(filePathOrUrl);
        } else {
          file = await readFile(filePathOrUrl);
          ext = extnameWithoutQuery(filePathOrUrl);
        }

        const filepath = `media/hash_${imagePathHash}${ext}`;

        zip.writeBinary(join("word", filepath), file);
        return xmlElementToJson(safeXml`
          <Relationship
            Id="${id}"
            Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/image"
            Target="${filepath}"
          />
        `);
      },
    ),
  );

  for (const element of elements) {
    if (element) {
      json.elements[0].elements.push(element);
    }
  }

  zip.writeJson(DOC_RELS, json as Element);
  return output;
}

export async function insertLinks<T extends Record<string, string>>(
  zip: JSZipCached,
  links: T,
) {
  const DOC_RELS = "word/_rels/document.xml.rels";
  const json = await zip.readJson(DOC_RELS);

  const existingIds = json.elements[0].elements
    .map((element: Element) => element.attributes?.Id)
    .filter(Boolean);
  const output: Record<keyof typeof links, string> = {} as Record<
    keyof typeof links,
    string
  >;

  const elements = Object.entries(links).map(([key, url]) => {
    const linkHash = generateHash(url);
    const id = "rId" + linkHash;
    output[key as keyof T] = id;

    if (existingIds.includes(id)) {
      return null;
    } else {
      existingIds.push(id);
    }

    return xmlElementToJson(safeXml`
      <Relationship
        Id="${id}"
        Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/hyperlink"
        Target="${url}"
        TargetMode="External"
      />
    `);
  });

  for (const element of elements) {
    if (element) {
      json.elements[0].elements.push(element);
    }
  }

  zip.writeJson(DOC_RELS, json as Element);
  return output;
}

async function addToContentTypesXml(zip: JSZipCached, xml: string) {
  const contentXml = await zip.readJson("[Content_Types].xml");
  contentXml.elements[0].elements.push(xmlElementToJson(xml));
  zip.writeJson("[Content_Types].xml", contentXml as Element);
}

async function insertHeaderFooters<T extends Record<string, string>>(
  type: "header" | "footer",
  zip: JSZipCached,
  content: T,
) {
  const DOC_RELS = "word/_rels/document.xml.rels";
  const json = await zip.readJson(DOC_RELS);

  const existingIds = json.elements[0].elements
    .map((element: Element) => element.attributes?.Id)
    .filter(Boolean);
  const output: Record<keyof typeof content, string> = {} as Record<
    keyof typeof content,
    string
  >;

  const elements = await Promise.all(
    Object.entries(content).map(async ([key, content]) => {
      const contentHash = generateHash(content).slice(0, 12);
      const id = "rId" + contentHash;
      output[key as keyof T] = id;

      if (existingIds.includes(id)) {
        return null;
      } else {
        existingIds.push(id);
      }

      const filepath = `header-${contentHash}.xml`;
      const zipfilepath = `word/${filepath}`;

      await addToContentTypesXml(
        zip,
        safeXml`
          <Override
            PartName="/${zipfilepath}"
            ContentType="application/vnd.openxmlformats-officedocument.wordprocessingml.${type}+xml"
          />
        `,
      );

      zip.writeString(zipfilepath, content.trim());

      return xmlElementToJson(`
        <Relationship Id="${id}" Type="http://purl.oclc.org/ooxml/officeDocument/relationships/${type}" Target="${filepath}" />
      `);
    }),
  );

  for (const element of elements) {
    if (element) {
      json.elements[0].elements.push(element);
    }
  }

  zip.writeJson(DOC_RELS, json as Element);
  return output;
}

export async function insertHeaders<T extends Record<string, string>>(
  zip: JSZipCached,
  content: T,
) {
  return insertHeaderFooters("header", zip, content);
}

export async function insertFooters<T extends Record<string, string>>(
  zip: JSZipCached,
  content: T,
) {
  return insertHeaderFooters("footer", zip, content);
}

export function cmToTwip(cm: number) {
  return Math.round(cm * 0.3937007874 * 1440);
}

// English Metric Unit (EMU): A measurement in computer typography. There are
// 635 EMUs per twip, 6,350 EMUs per half-point, 12,700 EMUs per point, and
// 914,400 EMUs per inch. These units are used to translate on-screen layouts
// to printed layouts for specified printer hardware.
export function cmToEmu(cm: number) {
  const inches = cm / 2.54;
  return Math.round(inches * 914400);
}

export function pointToDxa(input: number) {
  return input * 20;
}

export function cmToDxa(cm: number) {
  const inches = cm / 2.54;
  return inches * 72 * 20;
}

export function lineHeight(pointHeight: number, multiplier: number) {
  return Math.round(pointToDxa(pointHeight) * multiplier);
}

export function emuToCm(emu: number) {
  return (emu / 914400) * 2.54;
}

export function createHeader(id: string, type: string) {
  return safeXml`<w:headerReference r:id="${id}" w:type="${type}" />`;
}

export function createFooter(id: string, type: string) {
  return safeXml`<w:footerReference r:id="${id}" w:type="${type}" />`;
}

export function wrapInLinkTo(id: string, childXml: string) {
  return `<w:hyperlink r:id="${id}">${childXml}</w:hyperlink>`;
}

export function wrapInLinkToBookmark(anchor: string, childXml: string) {
  return `<w:hyperlink w:anchor="${anchor}">${childXml}</w:hyperlink>`;
}

export const CURRENT_BOOKMARK = { id: 10000 };
export function wrapInBookmarkPoint(anchor: string, childXml: string) {
  CURRENT_BOOKMARK.id++;
  return `<w:bookmarkStart w:id="${CURRENT_BOOKMARK.id}" w:name="${anchor}"/>${childXml}<w:bookmarkEnd w:id="${CURRENT_BOOKMARK.id}"/>`;
}

let IMAGE_ID = 10000;
type ImageOpts = {
  name?: string;
  desc?: string;
  width?: number;
  height?: number;
  xPos?: number;
  yPos?: number;
  xPosAnchor?: string;
  yPosAnchor?: string;
  isDecorative?: boolean;
  isWrapTight?: boolean;
  relativeHeight?: number;
};
export function createImage(rId: string, opts: ImageOpts = {}) {
  const uid = IMAGE_ID++;
  const {
    width = 1000000,
    height = 1000000,
    name = "",
    desc = "",
    xPos = undefined,
    yPos = undefined,
    xPosAnchor = "page",
    yPosAnchor = "page",
    isDecorative = false,
    isWrapTight = false,
    relativeHeight = 1,
  } = opts;

  const isDecorativeVal = isDecorative ? 1 : 0;

  const wrapInPosition = (xmlString: string) => {
    if (xPos === undefined && yPos === undefined) {
      return `<wp:inline distT="0" distB="0" distL="0" distR="0">${xmlString}</wp:inline>`;
    } else {
      return `<wp:anchor
        distT="0"
        distB="0"
        distL="0"
        distR="114300"
        simplePos="0"
        relativeHeight="${relativeHeight}"
        behindDoc="1"
        locked="0"
        layoutInCell="1"
        allowOverlap="1"
      >
        <wp:simplePos x="0" y="0"/>
        <wp:positionH relativeFrom="${xPosAnchor}">
            <wp:posOffset>${xPos}</wp:posOffset>
        </wp:positionH>
        <wp:positionV relativeFrom="${yPosAnchor}">
            <wp:posOffset>${yPos}</wp:posOffset>
        </wp:positionV>
        ${xmlString}
      </wp:anchor>`;
    }
  };

  return `
        <w:drawing>
            ${wrapInPosition(`
                <wp:extent cx="${width}" cy="${height}"/>
                <wp:effectExtent l="0" t="0" r="0" b="0"/>
                
                ${xPos && !isWrapTight ? "<wp:wrapNone />" : ""}
                ${
                  !isWrapTight
                    ? ""
                    : `
                  <wp:wrapTight wrapText="bothSides">
                    <wp:wrapPolygon edited="0" >
                      <wp:start x="0" y="0" />
                      <wp:lineTo x="0" y="0" />
                      <wp:lineTo x="0" y="0" />
                      <wp:lineTo x="0" y="0" />
                      <wp:lineTo x="0" y="0" />
                      <wp:lineTo x="0" y="0" />
                      <wp:lineTo x="0" y="0" />
                    </wp:wrapPolygon>
                  </wp:wrapTight>
                `
                }
                <wp:docPr id="${uid}" name="${name}" descr="${desc}">
                    <a:extLst xmlns:a="http://schemas.openxmlformats.org/drawingml/2006/main">
                        <a:ext uri="{C183D7F6-B498-43B3-948B-1728B52AA6E4}">
                            <adec:decorative xmlns:adec="http://schemas.microsoft.com/office/drawing/2017/decorative" val="${isDecorativeVal}"/>
                        </a:ext>
                    </a:extLst>
                </wp:docPr>
                <wp:cNvGraphicFramePr>
                    <a:graphicFrameLocks xmlns:a="http://schemas.openxmlformats.org/drawingml/2006/main" noChangeAspect="1"/>
                </wp:cNvGraphicFramePr>
                <a:graphic xmlns:a="http://schemas.openxmlformats.org/drawingml/2006/main">
                    <a:graphicData uri="http://schemas.openxmlformats.org/drawingml/2006/picture">
                        <pic:pic xmlns:pic="http://schemas.openxmlformats.org/drawingml/2006/picture">
                            <pic:nvPicPr>
                                <pic:cNvPr id="${uid}" name="${name}" descr="${desc}">
                                    <a:extLst>
                                        <a:ext uri="{C183D7F6-B498-43B3-948B-1728B52AA6E4}">
                                            <adec:decorative xmlns:adec="http://schemas.microsoft.com/office/drawing/2017/decorative" val="${isDecorativeVal}"/>
                                        </a:ext>
                                    </a:extLst>
                                </pic:cNvPr>
                                <pic:cNvPicPr/>
                            </pic:nvPicPr>
                            <pic:blipFill>
                                <a:blip r:embed="${rId}">
                                    <a:extLst>
                                        <a:ext uri="{28A0092B-C50C-407E-A947-70E740481C1C}">
                                            <a14:useLocalDpi xmlns:a14="http://schemas.microsoft.com/office/drawing/2010/main" val="0"/>
                                        </a:ext>
                                    </a:extLst>
                                </a:blip>
                                <a:stretch>
                                    <a:fillRect/>
                                </a:stretch>
                            </pic:blipFill>
                            <pic:spPr>
                                <a:xfrm>
                                    <a:off x="0" y="0"/>
                                    <a:ext cx="${width}" cy="${height}"/>
                                </a:xfrm>
                                <a:prstGeom prst="rect">
                                    <a:avLst/>
                                </a:prstGeom>
                            </pic:spPr>
                        </pic:pic>
                    </a:graphicData>
                </a:graphic>
            `)}
        </w:drawing>
    `;
}

export async function appendBodyElements(
  zip: JSZipCached,
  childElements: Element[] = [],
) {
  const doc = await zip.readJson("word/document.xml");
  if (!doc) {
    throw new Error("Missing ./word/document.xml");
  }

  const oldElements = doc.elements[0]!.elements![0].elements;
  doc.elements[0]!.elements![0].elements = [...oldElements, ...childElements];

  zip.writeJson("word/document.xml", doc as Element);
}

function notUndefined<TValue>(value: TValue | undefined): value is TValue {
  return value !== undefined;
}

export const checkWithinElement = (
  root: Element,
  fn: (el: Element, parent?: Element) => boolean,
  parent?: Element,
): boolean => {
  if (fn(root, parent)) {
    return true;
  }
  if (root.elements) {
    for (const element of root.elements) {
      if (checkWithinElement(element, fn, root)) {
        return true;
      }
    }
  }
  return false;
};

/**
 * Map over all elements from a root node
 */
export async function mapOverElements(
  root: Element,
  fn: (el: Element, parent?: Element) => Promise<Element | undefined>,
): Promise<Element | undefined> {
  const runner = async (
    root: Element,
    fn: (el: Element, parent?: Element) => Promise<Element | undefined>,
    parent?: Element,
  ): Promise<Element | undefined> => {
    const newRoot = await fn(root, parent);

    if (!newRoot) return;

    let hasChanged = false;
    if (newRoot.elements) {
      const newElementsSparse = await Promise.all(
        newRoot.elements.map(async (el) => {
          const moddedEl = await runner(el, fn, newRoot);
          if (moddedEl !== el) {
            hasChanged = true;
          }
          return moddedEl;
        }),
      );

      if (hasChanged) {
        return {
          ...newRoot,
          elements: newElementsSparse.filter(notUndefined),
        };
      }
    }

    return newRoot;
  };

  return runner(root, fn)!;
}

export class JSZipCached {
  public _zip: JSZip;
  private _cache: Record<string, Element | ElementCompact>;
  constructor(zip?: JSZip) {
    this._zip = zip ?? new JSZip();
    this._cache = {};
  }

  async readJson(filename: string) {
    if (this._cache[filename]) {
      return this._cache[filename]!;
    }
    return this._zip
      .file(filename)!
      .async("text")
      .then((res) => {
        const val = xmlRootToJson(res);
        this._cache[filename] = val;
        return val!;
      });
  }

  writeJson(filename: string, content: Element) {
    this._cache[filename] = content;
  }

  readBinary(filename: string) {
    return this._zip.file(filename)!.async("nodebuffer");
  }

  writeBinary(filename: string, content: Buffer) {
    return this._zip.file(filename, content);
  }

  readString(filename: string) {
    return this._zip.file(filename)!.async("string");
  }

  writeString(filename: string, content: string) {
    return this._zip.file(filename, content);
  }

  flush() {
    for (const [filename, content] of Object.entries(this._cache)) {
      this._zip.file(filename, jsonXmlToXmlString(content as Element));
    }
    this._cache = {};
  }

  getJsZip() {
    this.flush();
    return this._zip;
  }

  async zipToBuffer() {
    this.flush();
    return this._zip.generateAsync({
      type: "uint8array",
      mimeType:
        "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
      compression: "DEFLATE",
    });
  }
}

export async function generateEmptyDocx() {
  const basedir = join(
    process.cwd(),
    "./src/pages-helpers/curriculum/docx/empty-document.docx",
  );
  const files = await glob(`${basedir}/**/*`, { dot: true });

  const zip = new JSZip();
  for await (const file of files) {
    const zippath = relative(basedir, file);
    const res = await stat(file);
    if (res.isDirectory()) {
      zip.folder(zippath);
    } else {
      zip.file(zippath, await readFile(file));
    }
  }

  return new JSZipCached(zip);
}
