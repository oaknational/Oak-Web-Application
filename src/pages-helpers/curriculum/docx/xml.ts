import { js2xml, xml2js } from "xml-js";
import type { Element } from "xml-js";

export function xmlRootToJson(xmlData: string) {
  return xml2js(xmlData, {
    compact: false,
    captureSpacesBetweenElements: false,
  });
}

/**
 * To support <https://prettier.io/docs/en/options#embedded-language-formatting> as well as assertion.
 *
 * NOTE: This must called `xml`
 *
 * @param strings template strings
 * @param values tamplate values
 * @returns xml string
 */
export function safeXml(
  strings: TemplateStringsArray,
  ...values: (string | number | string[])[]
) {
  let outXml = "";
  for (let i = 0; i < strings.length; i++) {
    if (values[i] !== undefined) {
      const valueRaw = values[i];
      const value = Array.isArray(valueRaw) ? valueRaw.join("") : valueRaw;
      outXml += `${strings[i]!}${value}`;
    } else {
      outXml += `${strings[i]!}`;
    }
  }

  // Just for assertion
  xmlRootToJson(outXml);

  return outXml;
}

export function xmlElementToJson(xmlData?: string) {
  if (!xmlData) return;
  const rootNode = xmlRootToJson(xmlData);
  if (rootNode.elements.length !== 1) {
    throw new Error("Expecting a single root node in XML");
  }
  return collapseFragments(rootNode as Element)?.elements?.[0];
}

export function createFragment(elements: Element[]): Element {
  return {
    type: "element",
    name: "XML_FRAGMENT",
    elements: elements,
  };
}

export function collapseFragments(root: Element) {
  const run = (orig: Element) => {
    let node = orig;
    const cloneIfRequired = () => {
      if (node === orig) {
        return {
          ...node,
          elements: node.elements ? [...node.elements] : undefined,
        };
      }
      return node;
    };
    if (node.elements) {
      for (let i = node.elements.length - 1; i >= 0; i--) {
        if (node.elements) {
          const child = node.elements[i]!;
          const out = run(child);

          if (Array.isArray(out)) {
            node = cloneIfRequired();
            node.elements!.splice(i, 1, ...out);
          } else if (out !== child) {
            node = cloneIfRequired();
            if (node.elements) {
              node.elements[i] = out;
            }
          }
        }
      }
    }

    if (node.name === "XML_FRAGMENT") {
      return node.elements ?? [];
    }
    return node;
  };

  const out = run(root);
  if (Array.isArray(out)) {
    return createFragment(out);
  }
  return out;
}

export function jsonXmlToXmlString(json: Element): string {
  const root = "name" in json ? { elements: [json] } : json;
  const output = js2xml(collapseFragments(root));
  return output;
}

export function cdata(input: string | number) {
  return `<![CDATA[${input}]]>`;
}

export function cdataJson(input: Element): Element {
  if (input.type !== "text") {
    throw new Error("Expecting text node");
  }
  return {
    type: "cdata",
    cdata: String(input.text),
  };
}
