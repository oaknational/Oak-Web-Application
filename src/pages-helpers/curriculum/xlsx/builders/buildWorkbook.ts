import { safeXml } from "../../docx/xml";

export function buildWorkbook({ sheets }: { sheets: string[] }) {
  return safeXml`
    <?xml version="1.0" encoding="UTF-8" standalone="yes" ?>
    <workbook
      xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main"
      xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships"
      xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
      mc:Ignorable="x15 xr xr6 xr10 xr2"
      xmlns:x15="http://schemas.microsoft.com/office/spreadsheetml/2010/11/main"
      xmlns:xr="http://schemas.microsoft.com/office/spreadsheetml/2014/revision"
      xmlns:xr6="http://schemas.microsoft.com/office/spreadsheetml/2016/revision6"
      xmlns:xr10="http://schemas.microsoft.com/office/spreadsheetml/2016/revision10"
      xmlns:xr2="http://schemas.microsoft.com/office/spreadsheetml/2015/revision2"
    >
      <fileVersion
        appName="xl"
        lastEdited="7"
        lowestEdited="7"
        rupBuild="10511"
      />
      <workbookPr defaultThemeVersion="202300" />
      <mc:AlternateContent
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
      >
        <mc:Choice Requires="x15">
          <x15ac:absPath
            url="/Users/orangemug/Documents/"
            xmlns:x15ac="http://schemas.microsoft.com/office/spreadsheetml/2010/11/ac"
          />
        </mc:Choice>
      </mc:AlternateContent>
      <xr:revisionPtr
        revIDLastSave="0"
        documentId="8_{274615AA-ECE0-644D-AB2E-12D373184771}"
        xr6:coauthVersionLast="47"
        xr6:coauthVersionMax="47"
        xr10:uidLastSave="{00000000-0000-0000-0000-000000000000}"
      />
      <bookViews>
        <workbookView
          xWindow="1100"
          yWindow="820"
          windowWidth="28040"
          windowHeight="17440"
          xr2:uid="{C98B37CC-FD85-FE4F-984F-54449C3C0043}"
        />
      </bookViews>
      <sheets>${sheets}</sheets>
      <calcPr calcId="181029" />
      <extLst>
        <ext
          uri="{140A7094-0E35-4892-8432-C4D2E57EDEB5}"
          xmlns:x15="http://schemas.microsoft.com/office/spreadsheetml/2010/11/main"
        >
          <x15:workbookPr chartTrackingRefBase="1" />
        </ext>
        <ext
          uri="{B58B0392-4F1F-4190-BB64-5DF3571DCE5F}"
          xmlns:xcalcf="http://schemas.microsoft.com/office/spreadsheetml/2018/calcfeatures"
        >
          <xcalcf:calcFeatures>
            <xcalcf:feature name="microsoft.com:RD" />
            <xcalcf:feature name="microsoft.com:Single" />
            <xcalcf:feature name="microsoft.com:FV" />
            <xcalcf:feature name="microsoft.com:CNMTM" />
            <xcalcf:feature name="microsoft.com:LET_WF" />
            <xcalcf:feature name="microsoft.com:LAMBDA_WF" />
            <xcalcf:feature name="microsoft.com:ARRAYTEXT_WF" />
          </xcalcf:calcFeatures>
        </ext>
      </extLst>
    </workbook>
  `.trim();
}
