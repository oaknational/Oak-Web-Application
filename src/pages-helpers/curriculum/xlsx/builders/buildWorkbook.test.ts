import { safeXml } from "@ooxml-tools/xml";

import { prettyFormat } from "../../docx/xml";

import { buildWorkbook } from "./buildWorkbook";

describe("buildWorkbook", () => {
  test("with sheets", () => {
    const output = buildWorkbook({
      sheets: [
        safeXml`<stubSheet>STUB_1</stubSheet>`,
        safeXml`<stubSheet>STUB_2</stubSheet>`,
        safeXml`<stubSheet>STUB_3</stubSheet>`,
      ],
    });
    expect(prettyFormat(output)).toMatchSnapshot();
  });

  test("without sheets", () => {
    const output = buildWorkbook({
      sheets: [],
    });
    expect(prettyFormat(output)).toMatchSnapshot();
  });
});
