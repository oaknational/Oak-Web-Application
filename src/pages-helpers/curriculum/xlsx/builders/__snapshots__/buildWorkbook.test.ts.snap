// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildWorkbook with sheets 1`] = `
"<?xml
  version="1.0"
  encoding="UTF-8"
  standalone="yes"
?>
<workbook
  xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main"
  xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships"
  xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
  mc:Ignorable="x15 xr xr6 xr10 xr2"
  xmlns:x15="http://schemas.microsoft.com/office/spreadsheetml/2010/11/main"
  xmlns:xr="http://schemas.microsoft.com/office/spreadsheetml/2014/revision"
  xmlns:xr6="http://schemas.microsoft.com/office/spreadsheetml/2016/revision6"
  xmlns:xr10="http://schemas.microsoft.com/office/spreadsheetml/2016/revision10"
  xmlns:xr2="http://schemas.microsoft.com/office/spreadsheetml/2015/revision2"
>
  <fileVersion
    appName="xl"
    lastEdited="7"
    lowestEdited="7"
    rupBuild="10511"
  ></fileVersion>
  <workbookPr
    defaultThemeVersion="202300"
  ></workbookPr>
  <mc:AlternateContent
    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
  >
    <mc:Choice
      Requires="x15"
    >
      <x15ac:absPath
        url="/Users/orangemug/Documents/"
        xmlns:x15ac="http://schemas.microsoft.com/office/spreadsheetml/2010/11/ac"
      ></x15ac:absPath>
    </mc:Choice>
  </mc:AlternateContent>
  <xr:revisionPtr
    revIDLastSave="0"
    documentId="8_{274615AA-ECE0-644D-AB2E-12D373184771}"
    xr6:coauthVersionLast="47"
    xr6:coauthVersionMax="47"
    xr10:uidLastSave="{00000000-0000-0000-0000-000000000000}"
  ></xr:revisionPtr>
  <bookViews>
    <workbookView
      xWindow="1100"
      yWindow="820"
      windowWidth="28040"
      windowHeight="17440"
      xr2:uid="{C98B37CC-FD85-FE4F-984F-54449C3C0043}"
    ></workbookView>
  </bookViews>
  <sheets>
    <stubSheet>
      STUB_1
    </stubSheet>
    <stubSheet>
      STUB_2
    </stubSheet>
    <stubSheet>
      STUB_3
    </stubSheet>
  </sheets>
  <calcPr
    calcId="181029"
  ></calcPr>
  <extLst>
    <ext
      uri="{140A7094-0E35-4892-8432-C4D2E57EDEB5}"
      xmlns:x15="http://schemas.microsoft.com/office/spreadsheetml/2010/11/main"
    >
      <x15:workbookPr
        chartTrackingRefBase="1"
      ></x15:workbookPr>
    </ext>
    <ext
      uri="{B58B0392-4F1F-4190-BB64-5DF3571DCE5F}"
      xmlns:xcalcf="http://schemas.microsoft.com/office/spreadsheetml/2018/calcfeatures"
    >
      <xcalcf:calcFeatures>
        <xcalcf:feature
          name="microsoft.com:RD"
        ></xcalcf:feature>
        <xcalcf:feature
          name="microsoft.com:Single"
        ></xcalcf:feature>
        <xcalcf:feature
          name="microsoft.com:FV"
        ></xcalcf:feature>
        <xcalcf:feature
          name="microsoft.com:CNMTM"
        ></xcalcf:feature>
        <xcalcf:feature
          name="microsoft.com:LET_WF"
        ></xcalcf:feature>
        <xcalcf:feature
          name="microsoft.com:LAMBDA_WF"
        ></xcalcf:feature>
        <xcalcf:feature
          name="microsoft.com:ARRAYTEXT_WF"
        ></xcalcf:feature>
      </xcalcf:calcFeatures>
    </ext>
  </extLst>
</workbook>"
`;

exports[`buildWorkbook without sheets 1`] = `
"<?xml
  version="1.0"
  encoding="UTF-8"
  standalone="yes"
?>
<workbook
  xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main"
  xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships"
  xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
  mc:Ignorable="x15 xr xr6 xr10 xr2"
  xmlns:x15="http://schemas.microsoft.com/office/spreadsheetml/2010/11/main"
  xmlns:xr="http://schemas.microsoft.com/office/spreadsheetml/2014/revision"
  xmlns:xr6="http://schemas.microsoft.com/office/spreadsheetml/2016/revision6"
  xmlns:xr10="http://schemas.microsoft.com/office/spreadsheetml/2016/revision10"
  xmlns:xr2="http://schemas.microsoft.com/office/spreadsheetml/2015/revision2"
>
  <fileVersion
    appName="xl"
    lastEdited="7"
    lowestEdited="7"
    rupBuild="10511"
  ></fileVersion>
  <workbookPr
    defaultThemeVersion="202300"
  ></workbookPr>
  <mc:AlternateContent
    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
  >
    <mc:Choice
      Requires="x15"
    >
      <x15ac:absPath
        url="/Users/orangemug/Documents/"
        xmlns:x15ac="http://schemas.microsoft.com/office/spreadsheetml/2010/11/ac"
      ></x15ac:absPath>
    </mc:Choice>
  </mc:AlternateContent>
  <xr:revisionPtr
    revIDLastSave="0"
    documentId="8_{274615AA-ECE0-644D-AB2E-12D373184771}"
    xr6:coauthVersionLast="47"
    xr6:coauthVersionMax="47"
    xr10:uidLastSave="{00000000-0000-0000-0000-000000000000}"
  ></xr:revisionPtr>
  <bookViews>
    <workbookView
      xWindow="1100"
      yWindow="820"
      windowWidth="28040"
      windowHeight="17440"
      xr2:uid="{C98B37CC-FD85-FE4F-984F-54449C3C0043}"
    ></workbookView>
  </bookViews>
  <sheets></sheets>
  <calcPr
    calcId="181029"
  ></calcPr>
  <extLst>
    <ext
      uri="{140A7094-0E35-4892-8432-C4D2E57EDEB5}"
      xmlns:x15="http://schemas.microsoft.com/office/spreadsheetml/2010/11/main"
    >
      <x15:workbookPr
        chartTrackingRefBase="1"
      ></x15:workbookPr>
    </ext>
    <ext
      uri="{B58B0392-4F1F-4190-BB64-5DF3571DCE5F}"
      xmlns:xcalcf="http://schemas.microsoft.com/office/spreadsheetml/2018/calcfeatures"
    >
      <xcalcf:calcFeatures>
        <xcalcf:feature
          name="microsoft.com:RD"
        ></xcalcf:feature>
        <xcalcf:feature
          name="microsoft.com:Single"
        ></xcalcf:feature>
        <xcalcf:feature
          name="microsoft.com:FV"
        ></xcalcf:feature>
        <xcalcf:feature
          name="microsoft.com:CNMTM"
        ></xcalcf:feature>
        <xcalcf:feature
          name="microsoft.com:LET_WF"
        ></xcalcf:feature>
        <xcalcf:feature
          name="microsoft.com:LAMBDA_WF"
        ></xcalcf:feature>
        <xcalcf:feature
          name="microsoft.com:ARRAYTEXT_WF"
        ></xcalcf:feature>
      </xcalcf:calcFeatures>
    </ext>
  </extLst>
</workbook>"
`;
